internal/controller/bootstrap/controlplane_bootstrap_controller.go

```diff
@@ -0,0 +1,421 @@
+		return nil, err
+	}
+	files = append(files, certs...)
+
+	// Create the token using the child cluster client
+	tokenID := kutil.RandomString(6)
+	tokenSecret := kutil.RandomString(16)
+	token := fmt.Sprintf("%s.%s", tokenID, tokenSecret)
+	tokenKubeSecret := createTokenSecret(tokenID, tokenSecret)
+
+	chCS, err := c.getChildClusterClientSet(ctx, scope)
+	if err != nil {
+		log.Error(err, "Failed to getting child cluster client set")
+		return nil, err
+	}
+
+	_, err = chCS.CoreV1().Secrets("kube-system").Create(ctx, tokenKubeSecret, metav1.CreateOptions{})
+	if err != nil {
+		log.Error(err, "Failed to create token secret in the child cluster")
+		return nil, err
+	}
+
+	joinToken, err := kutil.CreateK0sJoinToken(ca.KeyPair.Cert, token, fmt.Sprintf("https://%s:%d", scope.Cluster.Spec.ControlPlaneEndpoint.Host, scope.Cluster.Spec.ControlPlaneEndpoint.Port))
+
+	files = append(files, cloudinit.File{
+		Path:        joinTokenFilePath,
+		Permissions: "0644",
+		Content:     joinToken,
+	})
+
+	return files, err
+}
+
+func (c *ControlPlaneController) getCerts(ctx context.Context, scope *Scope) ([]cloudinit.File, *secret.Certificate, error) {
+	var files []cloudinit.File
+	certificates := secret.NewCertificatesForInitialControlPlane(&kubeadmbootstrapv1.ClusterConfiguration{
+		CertificatesDir: "/var/lib/k0s/pki",
+	})
+
+	err := certificates.Lookup(ctx, c.Client, util.ObjectKey(scope.Cluster))
+	if err != nil {
+		return nil, nil, err
+	}
+	ca := certificates.GetByPurpose(secret.ClusterCA)
+	for _, cert := range certificates.AsFiles() {
+		files = append(files, cloudinit.File{
+			Path:        cert.Path,
+			Permissions: "0644",
+			Content:     cert.Content,
+		})
+	}
+
+	return files, ca, nil
+}
+
+func createTokenSecret(tokenID, tokenSecret string) *corev1.Secret {
+	return &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Namespace: "kube-system",
+		},
+		Type: corev1.SecretTypeBootstrapToken,
+		StringData: map[string]string{
+			"token-id":     tokenID,
+			"token-secret": tokenSecret,
+			// TODO We need bit shorter time for the token
+			"expiration":                       time.Now().Add(24 * time.Hour).Format(time.RFC3339),
+			"usage-bootstrap-api-auth":         "true",
+			"description":                      "Worker bootstrap token generated by k0smotron",
+			"usage-bootstrap-authentication":   "true",
+			"usage-bootstrap-api-worker-calls": "true",
+		},
+	}
+}
+
+func (c *ControlPlaneController) createKubeconfig(ctx context.Context, scope *Scope) error {
+	if scope.Cluster.Spec.ControlPlaneEndpoint.IsZero() {
+		return errors.New("control plane endpoint is not set")
+	}
+
+	return kubeconfig.CreateSecret(ctx, c.Client, scope.Cluster)
+}
+
+func (c *ControlPlaneController) getChildClusterClientSet(ctx context.Context, scope *Scope) (*kubernetes.Clientset, error) {
```

Comment:
https://pkg.go.dev/sigs.k8s.io/cluster-api@v1.5.0/controllers/remote#NewClusterClient

There's a generic helper for this

Reply:
Re-wrote it using the helper, thanks!